main.go

```diff
@@ -64,28 +61,22 @@ func main() {
 	go serveMetrics(cfg.MetricsAddress)
 	go handleSigterm(stopChan)
 
-	client, err := newClient(cfg)
-	if err != nil {
-		log.Fatal(err)
-	}
-
-	serviceSource, err := source.NewServiceSource(client, cfg.Namespace, cfg.FqdnTemplate, cfg.Compatibility)
-	if err != nil {
-		log.Fatal(err)
-	}
-	source.Register("service", serviceSource)
-
-	ingressSource, err := source.NewIngressSource(client, cfg.Namespace, cfg.FqdnTemplate)
-	if err != nil {
-		log.Fatal(err)
+	// Create a source.Config from the flags passed by the user.
+	sourceCfg := source.Config{
+		KubeMaster:    cfg.Master,
+		KubeConfig:    cfg.KubeConfig,
+		Namespace:     cfg.Namespace,
+		FQDNTemplate:  cfg.FQDNTemplate,
+		Compatibility: cfg.Compatibility,
 	}
-	source.Register("ingress", ingressSource)
 
-	sources, err := source.LookupMultiple(cfg.Sources)
+	// Lookup all the selected sources and pass them the desired configuration.
+	sources, err := source.LookupMultiple(cfg.Sources, &sourceCfg)
```

Comment:
Right, we should change NewFakeSource to copy cfg.FQDNTemplate rather than modifying it.

But, then we'd have to make kubeClient separate from cfg, so it can be shared by Ingress and Service sources.  Maybe even add a mutex around it?  Possibly premature, assuming there's no benefit in (future) concurrent instantiation of sources.

AIUI, go doesn't have immutable types.  Trusting contracts feels ... brittle.  (That's just a social convention, right, no way to enforce?)  So - copies for cfg, and split kubeClient out from cfg?